### DISCUSSION TITLE
Making Product Collection the main block for product lists and collections

### DISCUSSION THREAD

**Discussion by kmanijak** [woocommerce/woocommerce#44776] [WooCommerce, Blocks, and Site Editing]
# Background

There are many ways of rendering lists or collections of products across various blocks in WooCommerce. We can split them into two buckets:

- Generic ones:
  - All Products
  - Products (Beta)
  - Product Collection (Beta)
- Specialized ones:
  - Best Selling Products
  - Top Rated Products
  - Newest Products
  - On Sale Products
  - Hand-picked Products
  - Products by Category
  - Products by Tag
  - Products by Attribute

We put a lot of effort into making Product Collection (Beta) better than the other generic blocks and making it capable of covering all the cases offered by the specialized product list block. All of that with better, more modern UX and performance.

# Intended change

We want to make a Product Collection a go-to block to list your products, which means the following steps:

- Release the stable Product Collection block,
- Hide the other product blocks from the inserter,
- Use Product Collection as a default in product archive templates,
- In addition to Product Collection block, expose the “collections” that will substitute some blocks:
  - "Best Selling Products" will become "Top Sellers" collection,
  - "Top Rated Products" will become "Top Rated" collection,
  - "Newest Products" will become "New Arrivals" collection,
  - "On Sale Products" will become "On Sale" collection,
  - ...and more collections to come!

# What does it mean for you?

If a store already has existing product blocks in use, you may continue using those. However, when setting up a new product list, you will now see the Product Collection Block instead of the other product list blocks in the inserter.

**Before** | **After**
-- | --
Several blocks offer overlapping features and different interfaces | The Product Collection block along with more specific collections, under one unified interface.
![image](https://github.com/woocommerce/woocommerce/assets/20098064/7f1898e0-47e6-45d2-9749-6a621e31d2c7) | ![image](https://github.com/woocommerce/woocommerce/assets/20098064/e9f9fb5f-ea4f-42f9-a4da-1e5c2bcac0bc)

Additionally, all of the collections will be available when inserting Product Collection block:
![collection-inserter](https://github.com/woocommerce/woocommerce/assets/20098064/f8a65763-665d-43b5-9cfb-ea55760f271a)

# Extending the Product Collection Block

Product Collection is backward compatible with the Products (Beta) block in terms of extensibility. It offers the same level of extensibility as Products (Beta) block that is currently used as a default block in product archive templates. That means, however you extended the Products (Beta) blocks, you're able to achieve the same results with Product Collection (Beta) block.

### Frontend Hooks support

There's a Compatibility Layer available on the frontend for archive templates to extend. Compatibility Layer applies the same filters to Product Collection block as to Products (Beta):

- [Documentation](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/docs/internal-developers/blockified-templates/compatibility-layer.md)
- [Code reference](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/Templates/ArchiveProductTemplatesCompatibility.php)

In addition to the above, Product Collection uses the same inner blocks to build a product template (e.g. Product Price or Product Title). Nothing changes in terms of the extensibility of those elements.

### Plans

In addition to existing extensibility points, we're working on additional ones to allow 3rd Party Developers to create and register their own collections. Our goal is to enable you to prepare presets for Product Collection block that will predefine the product query and filters exposed to merchants in Editor.

Let's take the Top Rated collection as an example. It has the orderby attribute predefined and set to rating. In addition, there will be an option to define it as a hidden control, so the merchant cannot change it. However, they have a set of other Editor filters available to build on top of that. For example, they could apply category filter limiting the Top Rated collection to specific categories.

The next step would be allowing for creating and registering your own query filters that could be exposed in UI controls in Editor.

We're still working on it, but you can expect further communication once the first version is ready!

# Your feedback

Let us know what you think about that change. We’re particularly interested in:

- Is there any particular feature from the existing product list blocks that you feel is missing in the Product Collection block?
- Have you encountered any troubles extending the Product Collection block compared to Products (Beta) or other product list blocks?

Your feedback is valuable as we continue to refine and expand the capabilities of the Product Collection block.

**Comment by nerrad** [comment]
Seems like a solid plan to me. The soft deprecation you mentioned (removing from the inserter) will be important to ensure no existing usage of the blocks are broken. Some other things I think should be considered:

- updating patterns so they aren't using any of the deprecated blocks.
- some mechanism for clearly marking the deprecation in the code for the blocks. Eventually, it'd be nice to remove the code from the codebase (but it'd be a while before we'd be able to do that). Any signalling we can provide to developers early so they don't build things on top of that code will be important.
- updating any documentation (user or otherwise) that exists for the deprecated blocks.

**Comment by betayazilim** [comment]
Hi,
I use shortcodes on my woocommerce mainpage.
I tried to use product collections.
1. I can't limit featured products by category. I am trying to selected featured CPUs in one line, featured monitors in other line.
2. I can't shows stars/reviews in listing.
3. title, price and add to cart buttons are not in line, they shown in different height.

**Reply by kmanijak** [reply]
Hi @ned-bs , thanks for providing this feedback! It's super valuable, I hope you don't mind asking me some questions so we can better understand your use case.

> I can't limit featured products by category. I am trying to selected featured CPUs in one line, featured monitors in other line.

You can use Product Categories filter to display products from single category. I tried achieving that in the screenshot below. First collection displays featured products from one category and second one from another. Is that what you want to achieve?

![Screen Shot 2024-07-15 at 14 25 22 PM](https://github.com/user-attachments/assets/f4500761-f9d5-4f0e-b355-03263b877c06)

If yes, please follow these steps:
1. Add Product Collection block
2. Choose Featured collection (which only displays Featured products)
3. In the Inspector Controls, add Product Categories filter
4. Choose category you want (e.g. CPU)
5. Repeat 1-4 for another category (e.g. monitors)

And if that's not what you want to achieve, could you provide more details what did you try?

---

> I can't shows stars/reviews in listing.

You can use Product Rating block to display star rating along with all products. Did you have a chance to try it out?

<img width="986" alt="image" src="https://github.com/user-attachments/assets/b22c99fe-7a4c-4e74-9449-b557b9c6798c">


Reviews cannot be included in Product Collection block, that true.

> title, price and add to cart buttons are not in line, they shown in different height.

Do you mind sharing a screenshot to better explain what you mean?

**Reply by betayazilim** [reply]
Hi @kmanijak ,

Thanks for super fast reply.
1. I didn't realized that there are 3 dots there and can add more filters. I can filter by category now.  https://prnt.sc/8Ix1cvM6x6Fj 
2. I didn't know that I can add rating block between them. That is super cool. I have added.
3. What I mean is title's , reviews, add to cart's are not horizontally aligned. https://prnt.sc/wc9GuOcm5pSW

**Reply by kmanijak** [reply]
@ned-bs ,

> I didn't realized that there are 3 dots there and can add more filters.

Yeah, and there's even more filters you can play with and combine! :) 

> I didn't know that I can add rating block between them

There's more blocks that you can insert into Product Collection. You can also rearrange them if needed.

Here's [documentation](https://woocommerce.com/document/woocommerce-store-editing/customizing-shop-page-catalog/product-collection-block/) that provides some more information about the block's capabilities.

> What I mean is title's , reviews, add to cart's are not horizontally aligned. https://prnt.sc/wc9GuOcm5pSW

Ah, I see! That's a bug happening with WordPress 6.6. The fix is merged and will be available in WooCommerce 9.2:
https://github.com/woocommerce/woocommerce/pull/49096. And also, that bug appears only in the Editor and doesn't impact the customer-facing frontend:

Editor | Frontend
-- | --
<img width="1363" alt="image" src="https://github.com/user-attachments/assets/3eb34017-33e0-48a9-9576-b778563de8b0"> | <img width="1456" alt="image" src="https://github.com/user-attachments/assets/14845c51-f73f-4610-9870-15f6e90bca5f">

**Comment by betayazilim** [comment]
for #3: It was not I meant.
Images are inline in front end. Then there are titles, some of them long some of them short. So we haev 2 lines, 3 lines titles.
Then we have some products with reviews, some product without.
So that cause issues for rating and add to cart buttons. They are not in same line.
https://prnt.sc/sAtHnI57f3qH

**Reply by betayazilim** [reply]
I couldn't find sorting based on price in here. https://prnt.sc/sAtHnI57f3qH
I was able sort by price in shortcode.

**Comment by leewillis77** [comment]
I'm going to bring up an issue we raised at the beginning of the year that, as far as I'm aware a) still hasn't been addressed and b) presents a fundamental issue for any extension that outputs to the product listing pages. 

Currently, if an extension wants to output some content against a product on a product listing page they can trivially attach to a hook and create their output, for example:

`add_action( 'woocommerce_after_shop_loop_item_title', function() { echo 'FOO!'; }, 1, 0 );`

As I understand it though, even once we've migrated the content to a block, there's no way for a developer to automatically have their block added as an inner block in the product collection, and before a user can see the effect they would have to manually insert the block into any collection blocks on their site which is a way worse experience for the user.

Previous discussions on this: 
- https://github.com/WordPress/gutenberg/issues/54904#issuecomment-1875449258
- https://woocommercecommunity.slack.com/archives/C03MCL9RTAR/p1704283988088639

**Comment by kmanijak** [comment]
Hi @leewillis77, thanks for raising this topic again 🙌 

> I'm going to bring up an issue we raised at the beginning of the year that, as far as I'm aware a) still hasn't been addressed and b) presents a fundamental issue for any extension that outputs to the product listing pages.
> Currently, if an extension wants to output some content against a product on a product listing page they can trivially attach to a hook and create their output, for example:
> `add_action( 'woocommerce_after_shop_loop_item_title', function() { echo 'FOO!'; }, 1, 0 );`

Can I confirm that by "currently" you mean a shortcode or some other way of displaying products? I'm trying to make sure we're discussing the same use cases.

Also, another user raised a separate discussion on a very similar topic: https://github.com/woocommerce/woocommerce/discussions/52915.

**Comment by leewillis77** [comment]
> Can I confirm that by "currently" you mean a shortcode or some other way of displaying products? I'm trying to make sure we're discussing the same use cases.

I'm not restricting my comments to any particular way of displaying products - because it shouldn't matter *how* the product listing was added to the site. Let me try and explain the issue again (apologies if it's not clear!)

In a 'pre-blocks' world if a developer wants to write an extension that adds content to the product listing page then they can do so by attaching to a filter (woocommerce_after_shop_loop_item_title as an example, there are others depending where you want to add the content), and have their content added **automatically**, _wherever_ product listings are in use. 

If we fast forward to a 'post-blocks' world, as far as I can tell there's no way for a developer to have their block **automatically** inserted into _all_ product collection blocks. That is a significant regression in terms of extensability. Without that functionality, when the developer's extension is installed, nothing will change (from the users point of view), until they manually add the block to their product collection blocks one-by-one. That is an awful experience compared to the status quo.

As far as I can tell block hooks were (maybe) supposed to solve this, but from various discussions I've had - they actually don't.

So, the question is if I want to migrate an extension that currently adds content to the product listing pages via hooks over to using blocks, how do I make sure my block is added to product collection blocks by default? If that's not something that's currently supported, please can I speak loudly in favour of such a requirement being a strong blocker to considering the Product Collection block fit-for-purpose. 

> Also, another user raised a separate discussion on a very similar topic: https://github.com/woocommerce/woocommerce/discussions/52915.

This is a different issue as it relates to the availability of the hook backwards-compatibility layer which I'm not referring to at all.

**Reply by sunyatasattva** [reply]
> As far as I can tell block hooks were (maybe) supposed to solve this, but from various discussions I've had - they actually don't.

As I was reading this, it sounded to me exactly the use-case for block hooks. From the [introduction post](https://make.wordpress.org/core/2023/10/15/introducing-block-hooks-for-dynamic-blocks/):

> Front-end insertion should happen as soon as the plugin containing a hooked block is activated. In other words, the user isn’t required to insert the block in the Editor manually. Similarly, disabling the plugin should remove the hooked block from the front end.

Could you elaborate on why the block hooks didn't meet your requirements and perhaps link to the discussions you mentioned?

Like I mentioned in [this other thread](https://github.com/woocommerce/woocommerce/discussions/52915#discussioncomment-11315935), we are fully committed to a block-based future. So, while I agree with your sentiment that a plugin activation requiring extra user steps is a degraded experience on their perspective, we should work on a block-based solution instead of holding on to legacy paradigms.

For this specific use-case, I believe block hooks to be the solution. If they aren't, we need to figure out why and improve on them.

**Comment by leewillis77** [comment]
> Could you elaborate on why the block hooks didn't meet your requirements and perhaps link to the discussions you mentioned?

As far as I can tell it's not possible to have a block automatically added to all product collection blocks automatically. See the discussions I linked to previously:
- https://github.com/WordPress/gutenberg/issues/54904#issuecomment-1875449258
- https://woocommercecommunity.slack.com/archives/C03MCL9RTAR/p1704283988088639

If you believe that *is* possible, could you link to some documentation showing how it would be achieved?

**Reply by kmanijak** [reply]
Particular limitation is that Block Hooks API doesn't take effect in Post Content: https://core.trac.wordpress.org/ticket/61074.

I see and am happy to share there's some progress in Core in this area: https://github.com/WordPress/wordpress-develop/pull/7898. (the change could land in 6.8). Thank you @leewillis77 as you're one of the voices that highlighted the importance of it!

As @sunyatasattva said:

> Like I mentioned in https://github.com/woocommerce/woocommerce/discussions/52915#discussioncomment-11315935, we are fully committed to a block-based future. So, while I agree with your sentiment that a plugin activation requiring extra user steps is a degraded experience on their perspective, we should work on a block-based solution instead of holding on to legacy paradigms.
>
> For this specific use-case, I believe block hooks to be the solution. If they aren't, we need to figure out why and improve on them.

We want to rely on a Core's solution which in this case is Block Hooks API. Hence I'm planning to test and give feedback to the PR in Core to verify if it satisfies the needs. Feel free to do the same. 🙌

**Reply by leewillis77** [reply]
Thanks for the feedback @kmanijak!

I've tried the patch in https://github.com/WordPress/wordpress-develop/pull/7898, but it doesn't seem to resolve the issue we're seeing. Specifically, even with that patch in place hooked blocks aren't added to product collection blocks when included on pages.

Rather than post my exact code (which is a bit complex), I've pulled together a simple way of demonstrating the issue below.

Here's my setup:

1. Recent Woo install (9.4.2)
2. Following code in place (as a plugin / in an existing plugin / or in the theme's functions.php file):
```
add_filter( 'hooked_block_types', function($hooked_blocks, $position, $anchor_block, $context) {
    if ($anchor_block === 'woocommerce/product-price') {
        error_log('Received request to hook for anchor block woocommerce/product-price');
    }
    return $hooked_blocks;
}, 15, 4 );
```
3. Sample page created, with a "create your own" Product Collection block added.

If you visit the "Shop" page, you should see errors logged to the PHP error log indicating that there was a possibility to register hooked blocks for the woocommerce/product-price block, e.g.
```
[29-Nov-2024 10:41:03 UTC] Received request to hook for anchor block woocommerce/product-price
[29-Nov-2024 10:41:03 UTC] Received request to hook for anchor block woocommerce/product-price
[29-Nov-2024 10:41:03 UTC] Received request to hook for anchor block woocommerce/product-price
[29-Nov-2024 10:41:03 UTC] Received request to hook for anchor block woocommerce/product-price
```

If you visit the sample page that has the Product Collection block in place, then you'll see that no logs are generated, indicating that it was not possible to register hooked blocks for the woocommerce/product-price block.

**Reply by kmanijak** [reply]
Not sure why it may be. 🤔 I see you notified an author in the PR so let's wait for their response (but thank you for giving it a try!)

I'm using https://github.com/ockham/like-button as an example and here's the code I'm adding to `functions.php`. As you can see it attaches the blocks correctly for me in templates and posts 

1. Attaching block only on templates

```
function block_hooks_like_button($hooked_blocks, $position, $anchor_block, $context) {
	if ( $context instanceof WP_Block_Template ) {
		if (
			'woocommerce/product-price' === $anchor_block &&
			'before' === $position
		) {
			$hooked_blocks[] = 'ockham/like-button';
		}
	}

	return $hooked_blocks;
}

add_filter('hooked_block_types', 'block_hooks_like_button', 10, 4);
```

x | Editor | Frontend
-- | -- | --
Product Catalog - Like button is there | <img width="1162" alt="image" src="https://github.com/user-attachments/assets/97accf17-2ef4-4b30-be8b-da4a2f6b33f8"> | <img width="1277" alt="image" src="https://github.com/user-attachments/assets/cfbb9325-88e3-40ee-917c-5fe613ec8c7c">
Post - no Like button | <img width="722" alt="image" src="https://github.com/user-attachments/assets/2115590c-f788-4982-807a-1ee86344fd64"> | <img width="783" alt="image" src="https://github.com/user-attachments/assets/36fc6b7e-a7cc-43ef-a19a-d1d217e27558">


2. Attaching block only on posts

```
function block_hooks_like_button($hooked_blocks, $position, $anchor_block, $context) {
	if ( $context instanceof WP_Post ) {
		if (
			'woocommerce/product-price' === $anchor_block &&
			'before' === $position
		) {
			$hooked_blocks[] = 'ockham/like-button';
		}
	}

	return $hooked_blocks;
}

add_filter('hooked_block_types', 'block_hooks_like_button', 10, 4);
```

x | Editor | Frontend
-- | -- | --
Product Catalog - no like button | <img width="1173" alt="image" src="https://github.com/user-attachments/assets/1eda8288-a9dc-447b-8a62-646646d914bc"> | <img width="1279" alt="image" src="https://github.com/user-attachments/assets/6b86590e-a1a9-448e-a1aa-e6e64da048a6">
Post - Like button is there | <img width="723" alt="image" src="https://github.com/user-attachments/assets/2102e35b-50dd-443b-9db8-006d839bfa60"> | <img width="748" alt="image" src="https://github.com/user-attachments/assets/90ff9598-24d6-4a08-ae07-f1b79517f920">

3. Attaching block everywhere

```
function block_hooks_like_button($hooked_blocks, $position, $anchor_block, $context) {
	if (
		'woocommerce/product-price' === $anchor_block &&
		'before' === $position
	) {
		$hooked_blocks[] = 'ockham/like-button';
	}

	return $hooked_blocks;
}

add_filter('hooked_block_types', 'block_hooks_like_button', 10, 4);
```

x | Editor | Frontend
-- | -- | --
Product Catalog - Like button is there | <img width="1177" alt="image" src="https://github.com/user-attachments/assets/61e7d09a-06f1-455b-b0b0-ed286cc70798"> | <img width="1277" alt="image" src="https://github.com/user-attachments/assets/776cc785-c61f-43f8-b6e3-f2f0a3114c97">
Post - Like button is there | <img width="778" alt="image" src="https://github.com/user-attachments/assets/20ce4a5e-2f3f-4f76-aa43-a448bc37e0a9"> | <img width="830" alt="image" src="https://github.com/user-attachments/assets/bc33c562-2552-4a71-bf59-0ba8437102df">

**Reply by leewillis77** [reply]
Hmm. That's odd. I don't see that behaviour at all. I installed and activated https://github.com/ockham/like-button (and confirmed that the block is available for manual insertion).

I then copied your 3rd example into my theme's functions.php. I see the Like button hooked into the standard "Shop" display, but. it doesn't appear in Product Collection blocks in pages or posts.

I wonder if it's to do with how I've patched WP. I simply applied https://patch-diff.githubusercontent.com/raw/WordPress/wordpress-develop/pull/7898.diff to WP 6.7.1, but I'm wondering if there are other changes that I'm missing that are causing this. Are you testing on WP trunk? 6.7.1?

**Reply by leewillis77** [reply]
@kmanijak I've left a note on the core issue, and pinged you, but wanted to cover the woo-specific impact of those comments here. In the historic implementation there are separate hooks for the various product elements on the archive page, and single product display, e.g.

woocommerce_after_shop_loop_item_title
vs
woocommerce_single_product_summary

So, if we want to add something to the archive display or products, but not the single product page we can do so. The hooks here are conceptual 'sections' rather than actual product attributes - e.g. we're saying "add our output to the after_shop_loop_item_title section". 

For block hooks, there are no such positional sections, and all we can do is to hook blocks in relative to standard product elements, e.g. we have to say "add our output after the price block". This causes issue, because the block hooks mechanism doesn't provide any context about why a price is being output - ie we can no longer know if we're on a product archive (collection), a single product display, or even some completely custom set of blocks that a user has assembled. 

That makes the block hooks approach significantly less flexible that the historic hook approach, and for us doesn't deliver what we need to be able to make our extensions block-based.

**Comment by betayazilim** [comment]
Is there any way to sort products based on front end price?

**Comment by rnayabed** [comment]
Is the "Product Collection" block non-blocking? I ask because when using query monitor, i see the DB calls still increase considerably when using this. I cannot confirm because I could not understand much from looking at the block source code.

**Reply by kmanijak** [reply]
Could you elaborate more what do you mean by "non-blocking"?

**Reply by rnayabed** [reply]
Hi! By "non blocking" i meant if the block initially loads up with a spinner/animation and then populates it after site load with ajax calls, or if it works like the older "products by category" block, where it needs to be rendered by PHP (the rest of the site loading gets "blocked" until the block's content is rendered)

**Reply by rnayabed** [reply]
I have edited the older "product by category" block to the cache the HTML instead and show it directly, that decreased the DB queries by a lot. 

I was wondering if the newer block does something similar, or atleast relies on AJAX calls and populating the blocks space later, so that the site TTFB is maintained.

**Reply by kmanijak** [reply]
I see @gigitux responded to your questions in WooCommerce Community Slack, but let me iterate here.

> I don't think that it is something that we will work on soon [cache]. You could use an external plugin or implement your cache strategy on your own.

**Comment by rnayabed** [comment]
I am unable to get the same results as the old "products by category" block with the new "product collections" block (with "Product Categories" filter):

Old block:

![image](https://github.com/user-attachments/assets/a6146984-6aa5-4335-9ee4-123340110ae7)

New block:

![image](https://github.com/user-attachments/assets/d7da7539-de7b-4033-ba33-432b93b921ea)


The old block allowed us to select our own columns and rows, this new one does not. I am also unable to get the same output as the old block. How can we fix this?

The new block also has a minor bug where clicking on the block shows "product template". only clicking slightly above selects and opens the actual block (Product Collection):


https://github.com/user-attachments/assets/9032c3fd-339b-4112-a4bc-62a7d03fc1f7

**Reply by kmanijak** [reply]
Hey @rnayabed , thanks for the comments and feedback.

> I am unable to get the same results as the old "products by category" block with the new "product collections" block (with "Product Categories" filter)

To display all products by category, please choose "create your own":

<img width="673" alt="image" src="https://github.com/user-attachments/assets/05c91044-153d-4951-8048-8ccbe6ee739c" />

and you can control the number of products and columns with toolbar settings and inspector controls:

Number of products | Number of columns + responsiveness
-- | --
<img width="748" alt="image" src="https://github.com/user-attachments/assets/22fdba96-9844-46f8-b8b2-d424bc94a0c4" /> | <img width="272" alt="image" src="https://github.com/user-attachments/assets/c56dd6b4-1f23-410a-9f46-f503bce7a50d" />

From WooCommerce 9.6 these options will be placed next to each other in Inspector Controls:
<img width="274" alt="image" src="https://github.com/user-attachments/assets/f0c8b6c0-f1f9-4f8b-8f7b-e65116ac11e4" />

I hope that helps!

**Reply by rnayabed** [reply]
Thank you very much. I hope the UX is improved down the line, it seems to have taken a big hit compared the older block.

**Reply by kmanijak** [reply]
> I hope the UX is improved down the line, it seems to have taken a big hit compared the older block.

New block provides better flexibility and is more feature-rich. However, if you feel the UX downgraded, do you mind sharing more details? Is it about the focus aspect you described or there's more?

**Reply by rnayabed** [reply]
Sure! Some of the things I can note down:
- The old block allows us to easily change the rows columns. The new one, we need to play around with columns, and items per page, as you mentioned (happy to hear that the rows/cols thing will come back from WC 9.6)
- It also creates an extra "pagination" block that needs to be removed every time.
- The new block, as I mentioned previously (video attached above) has a little visual bug where we need to click slightly above the block to select the "product collection" block, otherwise a different (likely the underlying internal block) "product template" block is selected, which can be confusing
- We need to create our own collection (which I believe simply creates an empty? set of filters) if we want to feature products by category. Previously it was just a simple click (add the "products by category" block). If we select "Featured products" and then try adding the "Product categories" filter, we get the issue where the search results are inconsistent (fixed if i just create my own collection instead of selecting "featured products").
- The old block has a nice interface showing a nice tree of categories which we can select. the new one, we have to manually type, then wait for the autocomplete to finish, then select it from the dropdown. it seems like a visual regression from the older block.

While the new block is definitely much more flexible and allows more modification, I feel right now it's still a bit rough around the edges. The older block was a simple "search" > "add" and go! I can totally understand the additional steps in the new block due to it's generalised/flexible nature, however it would be good to have this as a "pre selected" option.

On a sidenote, I was going through it's source code (in order to add direct HTML caching) and realised its a beast compared to the older block (obvious reasons). Can you point to some resources/things I should know in order to be able to perform that? The older block had a simple `render` function in it's single php file, this one though is very generalised and i am not sure where to start with. Thank you very much for your time!

**Comment by saagasrns** [comment]
Feedback / experience with new Product Collection block: I´m using Extra Product Options, and when product has options, it crushes the Add to cart button. It is then one big balloon, which goes under the product below (!). Haven´t had this issue with any other blocks or former versions of product listings. Does anyone know how to fix this? Help needed 🙏

![Screenshot 2025-05-07 at 12-29-32 Kaikki tuotteet - Kolmas Silmä Puoti](https://github.com/user-attachments/assets/6d7fde52-6223-4e0e-bb49-3759a5c1639d)

**Reply by Aljullu** [reply]
Thanks for the report, @saagasrns! I was able to reproduce this issue as well. IMO, this needs to be fixed in Extra Product Options. They are using the `woocommerce_loop_add_to_cart_link` hook to print their custom markup for the button, but at the same time, they are printing the same CSS classes from the original button. So inside the _Product Collection_ block, they print a custom button layout but with the classes used by the _Product Collection_ block buttons. It's not strange that there are conflicts, as they are mixing a CSS prepared for one layout in a different layout. Unfortunately, there is little we can do from our side.

IMO, they way they should implement this is to set the products with extra product options as not supporting `ajax_add_to_cart` and, if needed, modify the button text with `woocommerce_product_add_to_cart_text`. There are extensions that handle this properly like [WooCommerce Product Add-Ons](https://woocommerce.com/products/product-add-ons/).

**Comment by rpldom** [comment]
How to extend product collection block?
I need it to show product short description and status by default.
I noticed this block uses under the hood ‘woocommerce\product-template’ hook.
So I wrote some code utilizing Blook Hooks API:

```
function example_block_hooks($hooked_blocks, $position, $anchor_block, $context) {

    if ($anchor_block === "woocommerce/product-template" && $position === 'before') {
        $hooked_blocks[] = 'woocommerce/product-summary';
        $hooked_blocks[] = 'woocommerce/product-stock-indicator';
    }


    return $hooked_blocks;
}

add_filter('hooked_block_types', 'example_block_hooks', 10, 4);
```

But it doesn’t work. When I add this code my markup gets ignoredHookedBlocks attribute. Why is that?

```
<div class="wp-block-woocommerce-product-collection">
<!-- wp:woocommerce/product-summary /-->
<!-- wp:woocommerce/product-stock-indicator /-->
<!-- wp:woocommerce/product-template {"metadata":{"ignoredHookedBlocks":["woocommerce/product-summary","woocommerce/product-stock-indicator"]}} -->
…
```

How would you extend product collection?
It’s so counter-intuitive and complex I’m loosing my mind.

**Reply by kmanijak** [reply]
Hi @Jeff1Jeff!

**Regsitering Custom Collection**

One option would be [Registering custom collections in product collection block](https://developer.woocommerce.com/docs/block-development/product-collection-block/register-product-collection/) where you can define a [custom inner blocks structure](https://developer.woocommerce.com/docs/block-development/product-collection-block/register-product-collection/#example-4-collection-with-inner-blocks) addressing your needs.

**Block Hooks**

If sticking with Block Hooks, you should not use Product template as an anchor since whatever's INSIDE of Product Template gets rendered. So basically you should choose some of the inner blocks you use inside a Product Template (for example Product Image or Product Price) and attach the Product Summary and Stock Indicator blocks to it.

**Future idea**

There's this idea to allow changing the default pattern used for Product Collection similar to Query Loop (https://github.com/woocommerce/woocommerce/issues/57943). It's not yet there but most likely we'll implement it.

**Last but not least - editor**

Obviously, it's not ideal and not super scalable but I need to mention the possibility to add the blocks you need from the editor.

I hope that helps you address your needs! 🙌

**Comment by leewillis77** [comment]
@kmanijak Block hooks aren't currently a usable solution, see https://core.trac.wordpress.org/ticket/63608

**Reply by ockham** [reply]
The behavior described in that ticket is expected (due to the anchor block being inserted _after_ plugin activation); I've left a [comment](https://core.trac.wordpress.org/ticket/63608#comment:4) on the ticket with further details. The _tl;dr_ is that if the anchor block is present _before_ the plugin is activated, Block Hooks should work as expected. I hope that this is an acceptable limitation 🙂

**Reply by leewillis77** [reply]
@ockham I've replied there, but as per my [comments earlier](https://github.com/woocommerce/woocommerce/discussions/44776#discussioncomment-11302618) in this thread, it's not I'm afraid.

**Comment by kmanijak** [comment]
I'm closing this discussion as Product Collection is a default for quite some time. 

Anyone, feel free to open dedicated issues or discussion about more specific topics where needed 🙌
